File: Data-Structures-Algorithms-In-TS/AlgorithmsChapters/DP/CoinChange.ts
function minCoins(coins: number[], n: number, val: number): number {
	if (val <= 0) return 0;

	let count = 0;
	coins.sort((a, b) => a - b);

	for (let i = n - 1; i >= 0 && val > 0;) {
		if (coins[i] <= val) {
			count++;
			val -= coins[i];
		} else {
			i--;
		}
	}
	return val === 0 ? count : -1;
}

function minCoins2(coins: number[], n: number, val: number): number {
	if (val === 0) return 0;

	let count = Infinity;
	for (let i = 0; i < n; i++) {
		if (coins[i] <= val) {
			const subCount = minCoins2(coins, n, val - coins[i]);
			if (subCount >= 0) count = Math.min(count, subCount + 1);
		}
	}
	return count !== Infinity ? count : -1;
}

function minCoinsTD(coins: number[], n: number, val: number): number {
	const count: number[] = new Array(val + 1).fill(Infinity);
	count[0] = 0;
	return minCoinsTDUtil(count, coins, n, val);
}

function minCoinsTDUtil(count: number[], coins: number[], n: number, val: number): number {
	if (count[val] !== Infinity) {
		return count[val];
	}

	for (let i = 0; i < n; i++) {
		if (coins[i] <= val) {
			const subCount = minCoinsTDUtil(count, coins, n, val - coins[i]);
			if (subCount !== Infinity && count[val] > subCount + 1) {
				count[val] = subCount + 1;
			}
		}
	}
	return count[val];
}

function minCoinsBU(coins: number[], n: number, val: number): number {
	const count: number[] = new Array(val + 1).fill(Infinity);
	count[0] = 0;

	for (let i = 1; i <= val; i++) {
		for (let j = 0; j < n; j++) {
			if (coins[j] <= i && count[i - coins[j]] !== Infinity && count[i] > count[i - coins[j]] + 1) {
				count[i] = count[i - coins[j]] + 1;
			}
		}
	}
	return count[val] !== Infinity ? count[val] : -1;
}

function printCoinsUtil(cvalue: number[], val: number): string {
	var str = "";
	if (val > 0) {
		str = printCoinsUtil(cvalue, val - cvalue[val]);
		str += `${cvalue[val]} `;
	}
	return str;
}

function printCoins(cvalue: number[], val: number): void {
	var str = "Coins are: ";
	str += printCoinsUtil(cvalue, val);
	console.log(str);
}

function minCoinsBU2(coins: number[], n: number, val: number): number {
	const count: number[] = new Array(val + 1).fill(Infinity);
	const cvalue: number[] = new Array(val + 1).fill(Infinity);

	count[0] = 0;

	for (let i = 1; i <= val; i++) {
		for (let j = 0; j < n; j++) {
			if (coins[j] <= i && count[i - coins[j]] !== Infinity && count[i] > count[i - coins[j]] + 1) {
				count[i] = count[i - coins[j]] + 1;
				cvalue[i] = coins[j];
			}
		}
	}

	if (count[val] === Infinity) return -1;

	printCoins(cvalue, val);
	return count[val];
}

// Test example
const coins: number[] = [1, 2, 5, 10];
const value: number = 16;
const n: number = coins.length;

console.log("Count is:", minCoins(coins, n, value));
console.log("Count is:", minCoins2(coins, n, value));
console.log("Count is:", minCoinsTD(coins, n, value));
console.log("Count is:", minCoinsBU(coins, n, value));
console.log("Count is:", minCoinsBU2(coins, n, value));

/*
Count is: 3
Count is: 3
Count is: 3
Count is: 3
Coins are: 10 5 1 
Count is: 3
*/

File: Data-Structures-and-Algorithms-using-Swift/AlgorithmsChapters/DP/CoinChange.swift
import Foundation

// Greedy may be wrong.
func minCoins(_ coins : inout [Int], _ n : Int, _ value : Int) -> Int {
	var val : Int = value
	if (val <= 0) {
		return 0
	}
	var count : Int = 0
	coins = coins.sorted()
	var i : Int = n - 1
	while (i >= 0 && val > 0) {
		if (coins[i] <= val) {
			count += 1
			val -= coins[i]
		} else {
			i -= 1
		}
	}
	return (val == 0) ? count : -1
}

// Brute force.
func minCoins2(_ coins : inout [Int], _ n : Int, _ value : Int) -> Int {
	let val : Int = value
	if (val == 0) {
		return 0
	}
	var count : Int = Int.max
	var i : Int = 0
	while (i < n) {
		if (coins[i] <= val) {
			let subCount : Int = minCoins2( &coins,n,val - coins[i])
			if (subCount >= 0) {
				count = min(count,subCount + 1)
			}
		}
		i += 1
	}
	return (count != Int.max) ? count : -1
}

func minCoinsTD(_ coins : inout [Int], _ n : Int, _ val : Int) -> Int {
	var count : [Int] = Array(repeating: Int.max, count: val + 1)
	count[0] = 0 // zero val need zero coins.
	return minCoinsTD( &count, &coins,n,val)
}

func minCoinsTD(_ count : inout [Int], _ coins : inout [Int], _ n : Int, _ val : Int) -> Int {
	// Base Case
	if (count[val] != Int.max) {
		return count[val]
	}

	// Recursion
	var i : Int = 0
	while (i < n) {
		if (coins[i] <= val) { // For all possible coins
			// check validity of a sub-problem
			let subCount : Int = minCoinsTD( &count, &coins,n,val - coins[i])
			if (subCount != Int.max && count[val] > (subCount + 1)) {
				count[val] = subCount + 1
			}
		}
		i += 1
	}
	return count[val]
}
func minCoinsBU(_ coins : inout [Int], _ n : Int, _ val : Int) -> Int {
	// DP bottom up approach.
	var count : [Int] = Array(repeating: Int.max, count: val + 1)
	count[0] = 0

	// Base value.
	var i : Int = 1
	while (i <= val) {
		var j : Int = 0
		while (j < n) {
			// For all coins smaller than or equal to i.
			if (coins[j] <= i && count[i - coins[j]] != Int.max && count[i] > count[i - coins[j]] + 1) {
				count[i] = count[i - coins[j]] + 1
			}
			j += 1
		}
		i += 1
	}
	return (count[val] != Int.max) ? count[val] : -1
}

func printCoinsUtil(_ cvalue : inout [Int], _ val : Int) {
	if (val > 0) {
		printCoinsUtil( &cvalue,val - cvalue[val])
		print(String(cvalue[val]), terminator: " ")
	}
}

func printCoins(_ cvalue : inout [Int], _ val : Int) {
	print("Coins are : ",terminator: "")
	printCoinsUtil( &cvalue,val)
	print()
}

// DP bottom up approach.
func minCoinsBU2(_ coins : inout [Int], _ n : Int, _ val : Int) -> Int {
	var count : [Int] = Array(repeating: Int.max, count: val + 1)
	var cvalue : [Int] = Array(repeating: Int.max, count: val + 1)
	count[0] = 0 // Base value.
	var i : Int = 1
	while (i <= val) {
		var j : Int = 0
		while (j < n) {
			// For all coins smaller than or equal to i.
			if (coins[j] <= i && count[i - coins[j]] != Int.max && count[i] > count[i - coins[j]] + 1) {
				count[i] = count[i - coins[j]] + 1
				cvalue[i] = coins[j]
			}
			j += 1
		}	
		i += 1
	}
	
	if (count[val] == Int.max) {
		return -1
	}
	printCoins( &cvalue,val)
	return count[val]
}

// Testing code.
var coins : [Int] = [5, 6]
let value : Int = 16
let n : Int = coins.count
print("Count is : " + String(minCoins( &coins,n,value)))
print("Count is : " + String(minCoins2( &coins,n,value)))
print("Count is : " + String(minCoinsTD( &coins,n,value)))
print("Count is : " + String(minCoinsBU( &coins,n,value)))
print("Count is : " + String(minCoinsBU2( &coins,n,value)))

/* 
Count is : -1
Count is : 3
Count is : 3
Count is : 3
Coins are : 6 5 5 
Count is : 3
 */